libpolyml/elfexport.cpp

```diff
@@ -118,6 +118,133 @@
+# define HOST_DIRECT_FPTR_RELOC R_386_32
+# define USE_RELA 0
+#elif defined(HOSTARCHITECTURE_PPC)
+# define HOST_E_MACHINE EM_PPC
+# define HOST_DIRECT_DATA_RELOC R_PPC_ADDR32
+# define HOST_DIRECT_FPTR_RELOC R_PPC_ADDR32
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_PPC64)
+# define HOST_E_MACHINE EM_PPC64
+# define HOST_DIRECT_DATA_RELOC R_PPC64_ADDR64
+# define HOST_DIRECT_FPTR_RELOC R_PPC64_ADDR64
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_S390)
+# define HOST_E_MACHINE EM_S390
+# define HOST_DIRECT_DATA_RELOC R_390_32
+# define HOST_DIRECT_FPTR_RELOC R_390_32
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_S390X)
+# define HOST_E_MACHINE EM_S390
+# define HOST_DIRECT_DATA_RELOC R_390_64
+# define HOST_DIRECT_FPTR_RELOC R_390_64
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_SH)
+# define HOST_E_MACHINE EM_SH
+# define HOST_DIRECT_DATA_RELOC R_SH_DIR32
+# define HOST_DIRECT_FPTR_RELOC R_SH_DIR32
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_SPARC)
+# define HOST_E_MACHINE EM_SPARC
+# define HOST_DIRECT_DATA_RELOC R_SPARC_32
+# define HOST_DIRECT_FPTR_RELOC R_SPARC_32
+# define USE_RELA 1
+/* Sparc/Solaris, at least 2.8, requires ELF32_Rela relocations.  For some reason,
+   though, it adds the value in the location being relocated (as with ELF32_Rel
+   relocations) as well as the addend. To be safe, whenever we use an ELF32_Rela
+   relocation we always zero the location to be relocated. */
+#elif defined(HOSTARCHITECTURE_SPARC64)
+# define HOST_E_MACHINE EM_SPARCV9
+# define HOST_DIRECT_DATA_RELOC R_SPARC_64
+# define HOST_DIRECT_FPTR_RELOC R_SPARC_64
+/* Use the most relaxed memory model. At link time, the most restrictive one is
+   chosen, so it does no harm to be as permissive as possible here. */
+# define HOST_E_FLAGS EF_SPARCV9_RMO
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_X86_64)
+/* It seems Solaris/X86-64 only supports ELF64_Rela relocations.  It appears that
+   Linux will support either so we now use Rela on X86-64. */
+# define HOST_E_MACHINE EM_X86_64
+# define HOST_DIRECT_DATA_RELOC R_X86_64_64
+# define HOST_DIRECT_FPTR_RELOC R_X86_64_64
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_X32)
+# define HOST_E_MACHINE EM_X86_64
+# define HOST_DIRECT_DATA_RELOC R_X86_64_32
+# define HOST_DIRECT_FPTR_RELOC R_X86_64_32
+# define USE_RELA 1
+#elif defined(HOSTARCHITECTURE_ARM)
+# ifndef EF_ARM_EABI_VER4
+#  define EF_ARM_EABI_VER4     0x04000000
+# endif
+// When linking ARM binaries the linker checks the ABI version.  We
+// need to set the version to the same as the libraries.
+// GCC currently uses version 4.
+# define HOST_E_MACHINE EM_ARM
+# define HOST_DIRECT_DATA_RELOC R_ARM_ABS32
+# define HOST_DIRECT_FPTR_RELOC R_ARM_ABS32
+# define USE_RELA 0
+# define HOST_E_FLAGS EF_ARM_EABI_VER4
+#elif defined(HOSTARCHITECTURE_HPPA)
+# if defined(__hpux)
+#  define HOST_OSABI ELFOSABI_HPUX
+# elif defined(__NetBSD__)
+#  define HOST_OSABI ELFOSABI_NETBSD
```

Comment:
This isn't correct. NetBSD sets `ELFOSABI_SYSV`.

Reply:
Could you submit a pull request for this.  I know it's a trivial change but I don't want to change something unless it's actually been tested by someone who has tried NetBSD on HPPA.

Reply:
Are you sure? I just downloaded the NetBSD-7.1-hppa.iso, mounted it, and extracted usr/lib/libarchive.a from HPPA/BINARY/SETS/COMP.TGZ, which readelf shows as:

```
File: usr/lib/libarchive.a(archive_write_set_format_by_name.o)
ELF Header:
  Magic:   7f 45 4c 46 01 02 01 02 00 00 00 00 00 00 00 00 
  Class:                             ELF32
  Data:                              2's complement, big endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - NetBSD
  ABI Version:                       0
  Type:                              REL (Relocatable file)
  Machine:                           HPPA
  Version:                           0x1
  Entry point address:               0x0
  Start of program headers:          0 (bytes into file)
  Start of section headers:          700 (bytes into file)
  Flags:                             0x210, PA-RISC 1.1
  Size of this header:               52 (bytes)
  Size of program headers:           0 (bytes)
  Number of program headers:         0
  Size of section headers:           40 (bytes)
  Number of section headers:         16
  Section header string table index: 13

[...]
```

Reply:
Hmm, I will recheck it, but anyway we don't check OSABI so this is not a big deal.